devito/types/object.py

```diff
@@ -84,6 +87,124 @@ def function(self):
     __reduce_ex__ = Pickable.__reduce_ex__
 
 
+class AbstractObjectWithShape(Basic, sympy.Basic, Pickable):
+
+    """
+    Base class for objects with derived type.
+
+    The hierarchy is structured as follows
+
+                         AbstractObjectWithShape
+                                   |
+                              PetscObject
```

Comment:
a PetscObject is way too special to live inside the `types/object` directory. Only the more general class belong here. 

The PetscObject type should instead be defined wherever you will implement the petsc machinery (`passes/iet/petsc/` ?) . Just like we do for some MPI Objects under `/mpi/distributed` or `/mpi/routines`